### DISCUSSION TITLE
server (docker) restart to keep test types runs isolated

### DISCUSSION THREAD

**Discussion by franz1981** [TechEmpower/FrameworkBenchmarks#7358] [Ideas]
Reading the documentation it seems that the server (docker) istance is kept running across different test types (eg db, update, json, plaintext,...).
Both native and JIT technology stacks can have trouble to keep performance stable across different runs of very different test types:
- native frameworks because of defragmentation of default allocators (if used)
- JIT ones due to deoptimizations and increasing pollution of runtime info due to the different workload nature

To keep results between types more stable would be better to restart the server (docker) machine before each validation/warmup/prime run, in order to have repeatable results independent by the order of test types, wdyt?

**Comment by franz1981** [comment]
@nbrady-techempower
Just to help giving a context to my request here;

Let me show you the numbers of running a branch of techempower with server container's restarted (naively implemented, but it's just to show the difference) vs master.

Just as reminder , the order of execution of tests:
- DB
- JSON
- QUERY
- UPDATE
- FORTUNE
- PLAINTEXT

The plaintext one, given  that's the last one and due to how the JVM compilation works, can have a huge variance in its results because of the previous executed (very different) workloads, see:

![image](https://user-images.githubusercontent.com/13125299/170660666-94cb7278-9618-4750-8525-cfa46349936b.png)

these results coming from our CI (that's using bare metal/idle/stable hw):
- 8150 is using techempower master
- 8193 is using [restarting docker for each test type](https://github.com/franz1981/FrameworkBenchmarks/commit/338ce118665ca4e4e9e71ed3152e98974ed8dae7) 

If you have a CI there, you can reproduce yourself just by running a plaintext alone vs *all* and checking plaintext results.
I have tried with others Java frameworks, getting very similar results (but I let the other frameworks authors to notice and comment).

**Reply by franz1981** [reply]
@nbrady-techempower any news on this?
I am currently on PTO but I can send a pr on my back or open the discussion elsewhere to collect opinions, if needed.
Very likely other JVM-based frameworks running different test types suffer from the same issue

**Comment by billywhizz** [comment]
i think it is better to re-use the existing server for all tests as it currently does - this is the intended purpose of the benchmarks afaict - to test frameworks across a wide range of different tasks, not to see how optimized they can be for each specific task. there is at least one framework already that re-compiles itself based on profiling information so it can better fit the expected load of the test. imho, this tends to obscure any advantage dynamic/JIT languages have over statically compiled ones, which should be something we can learn from these benchmarks. :man_shrugging:

**Reply by franz1981** [reply]
Re
> i think it is better to re-use the existing server for all tests as it currently does

Not sure is what a benchmark would look like, unless the purpose of the benchmark is to stress how good the JIT is to deoptimize and recompile code too (and simpler frameworks obviously are going to benefit by this ^^).
I agree about keeping the same static project and config for all benchmarks types, but not the same runtime server, especially If some code paths are shared across test runs.
The purpose of a benchmark is to get repeatable results and having results that vary depending of the order of tests doesn't seem to adhere to this principle.

If the purpose of the benchmark, instead is to emulate realistic scenarios we should:
- have a single server configuration for each framework 
- stress all endpoints with different type tests in parallel (and with different rate and/or load distributions)
- avoid aggressive pipelining because in real world this is not likely to happen, really

**Reply by franz1981** [reply]
Just noticed that https://github.com/TechEmpower/FrameworkBenchmarks/issues/7475#issuecomment-1191789666 mentioned about this as well

**Reply by billywhizz** [reply]
i agree with your suggestions re. emulating realistic scenarios! to be fair i think there has been much confusion about the purpose of the TE benchmarks but the TE team has made it quite clear that the primary purpose is to provide a "level playing field" and give a picture of relative performance of frameworks *across a range* of different scenarios/loads. personally, i would like to be able to see more techniques for max performance but i'm just not sure this fits with the stated objectives of the TE team.

re. your JIT issues outlined above - that sounds like very odd behaviour from the jvm - a 2x performance increase for JIT code fitted to a specific test/route? how realistic is it that you would ever build something like this out in the world? surely you want to see how your framework/server/platform works across a variety of different loads rather than having to build a specific server and have the JIT engine optimize if for a single route only? :man_shrugging:

**Reply by franz1981** [reply]
> the primary purpose is to provide a "level playing field" and give a picture of relative performance of frameworks across a range of different scenarios/loads

Yep but that doesn't mean that the same runtime env (always on) must be reused *across* the range of loads, because it won't make results reproducible (unless proven differently) depending by duration and order of tests, regardless the issue we're hitting for the quarkus framework.

> re. your JIT issues outlined above - that sounds like very odd behaviour from the jvm - a 2x performance increase for JIT code fitted to a specific test/route? 

I agree - it seems a kind of overfitting compilation issue on the JIT due to validation runs order making the code to compile in a way that fit very badly with network-limited use cases

> how realistic is it that you would ever build something like this out in the world?

From this pov I don't think the all-out-throughput with single workload running for X seconds with Y pipeline level to be realistic as well 
(see http://highscalability.com/blog/2015/10/5/your-load-generator-is-probably-lying-to-you-take-the-red-pi.html for more info why `wrk` is a bad tool for the job - and wrk2 would be way better in this), but I believe this has to be a separate issue to be addressed for this benchmark.

Isolation is a requirement for reproducibility from any benchmark (as per any scientific experiment); if we give up on this, it's fine, but I would like to understand why and how it relates with the nature of the TE benchmark - although native compiled applications (not using PGO, for example) won't benefit from a compiled code point of view, they would benefit from a reduced memory fragmentation, while GC-backed applications (assuming a copy/moving/compacting GC) won't get any benefit; it's a trade-off that would benefit differently both application types, while keeping results more repeatable. 

Said that, I agree that the issue we're seeing is pretty weird, but I can tell quarkus isn't the sole affected Java framework; I invite the other authors to share their opinions if they thing it worth to be clarified :)
Thanks for your time sharing your thoughts @billywhizz , I've appreciated !

**Reply by Sanne** [reply]
> re. your JIT issues outlined above - that sounds like very odd behaviour from the jvm - a 2x performance increase for JIT code fitted to a specific test/route? how realistic is it that you would ever build something like this out in the world? surely you want to see how your framework/server/platform works across a variety of different loads rather than having to build a specific server and have the JIT engine optimize if for a single route only? man_shrugging

Absolutely, and that's part of the problem: it's extremely unrealistic for a server application to be deployed and then be used for some time perfoming exclusively workloads of type A, and then being repurposed (while running the same code) to run exclusively workloads of type B. 

In a "normal" context, the same application would be handling requests of A anb B concurrently, or at least partially in overlap.

**Comment by joanhey** [comment]
But now there are fws than use a different variant per test, and benefiting from the JIT.

So we need to clarify the rules.
All frameworks can do it or not ?

**Reply by franz1981** [reply]
@billywhizz @joanhey let me report this as conclusion of our journey to fix the weird JIT behaviour we have seen 
https://redhatperf.github.io/post/type-check-scalability-issue/

And funny enough, it's very likely that many frameworks are affected ;) have fun!
